### ISSUE TITLE
Custom abstract generic sub-classes

### ISSUE THREAD

**Issue by chadrik** [python/mypy#2291]
The second class definition fails type-checking with `Invalid type "mymodule.T"`:

``` python
T = TypeVar('T')

class MyAbstractGeneric(Generic[T]):
    pass

class MyGeneric(MyAbstractGeneric[T]):
    pass
```

There's no documentation on creating custom abstract generic classes, and mypy and the typing tests have no examples of this either.  The most obvious examples are those in `typing` itself, but it apparently uses techniques that are off limits outside of `typing`.  I would love to know how to use multiple levels of inheritance to build up custom generics.

This is related to #606 but I did not want to derail the discussion there as my issue is more focused on creating generics rather than aliasing them.  Maybe just user error.

**Comment by JukkaL** [comment]
You can write it like this:

```
T = TypeVar('T')

class MyAbstractGeneric(Generic[T]):
    pass

class MyGeneric(MyAbstractGeneric[T], Generic[T]):
    pass
```

You have to declare the type variables using a `Generic[...]` base class, even if another base class is generic. There is an open issue for fixing this (#302).

Would you like to contribute a fix (either a fix to #302 or a documentation update)?

**Comment by JukkaL** [comment]
Leaving this open as a missing documentation issue.

**Comment by chadrik** [comment]
@JukkaL thanks for the fast response.  I think I can make time for the documentation update, but I'll probably wait until I've answered a few more questions for myself.

**Comment by gvanrossum** [comment]
Um, PEP 484 is explicit that this should be allowed:

> The ``Generic[T]`` base class is redundant in simple cases where you
> subclass some other generic class and specify type variables for its
> parameters
> ```python
>   from typing import TypeVar, Iterator
> 
>   T = TypeVar('T')
> 
>   class MyIter(Iterator[T]):
>       ...
> ```
> That class definition is equivalent to:
> ```python
>   class MyIter(Iterator[T], Generic[T]):
>       ...
> ```

So I rather think that we should fix this instead of documenting it. (Though I'm okay with the fix being scheduled for Milestone 0.5 and the docs updated temporarily to explain the issue.)

**Comment by JukkaL** [comment]
Yeah, I was aware of this -- #302 that I referred to above is about this.
